agent/containers.go

```diff
@@ -0,0 +1,91 @@
+package agent
+
+//go:generate mockgen -destination ./containers_mock.go -package agent . ContainerLister
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/httpapi"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/quartz"
+)
+
+const (
+	defaultGetContainersCacheDuration = 10 * time.Second
+	dockerCreatedAtTimeFormat         = "2006-01-02 15:04:05 -0700 MST"
+	getContainersTimeout              = 5 * time.Second
+)
+
+type containersHandler struct {
+	cacheDuration time.Duration
+	cl            ContainerLister
+	clock         quartz.Clock
+
+	mu         sync.Mutex // protects the below
+	containers []codersdk.WorkspaceAgentContainer
+	mtime      time.Time
+}
+
+func (ch *containersHandler) handler(rw http.ResponseWriter, r *http.Request) {
+	ct, err := ch.getContainers(r.Context())
+	if err != nil {
+		httpapi.Write(r.Context(), rw, http.StatusInternalServerError, codersdk.Response{
```

Comment:
For later: we should consider adding more specific error handling here
i.e. if the docker CLI can't be located, permissions error, etc
We could then send back specific codes to the calling API to provide a clearer error explanation & links to docs for resolution